peps/pep-0735.rst

```diff
@@ -0,0 +1,613 @@
+Path Dependencies.
+
+If the ``path`` values are distinct, but refer to the same concrete file or
+directory on the filesystem, handling behavior is unspecified. Implementations
+MAY normalize these paths to a single value.
+
+If the ``path`` values are identical, implementations MUST treat the result as
+a singular Path Dependency following the rules below:
+
+- The ``extras`` keys are concatenated into a single list
+- If ``editable`` is always ``true`` or always ``false``, it is treated as
+  such. Otherwise, the ``editable`` key is treated as though it were absent.
+- If the ``only_deps`` key is present, a ``false`` or absent value takes
+  priority over any ``true`` values. In other words, implementations respect
+  ``only_deps`` if it is ``true`` for all instances of the same ``path``.
+  Otherwise, they treat it as ``false``.
+
+
+Example Dependency Groups Table
+-------------------------------
+
+The following is an example of a partial ``pyproject.toml`` which uses this to
+define four dependency groups: ``test``, ``docs``, ``typing``, and
+``typing-test``:
+
+.. code:: toml
+
+    [dependency-groups]
+    test = ["pytest", "coverage", {path = ".", editable = true}]
+    docs = ["sphinx", "sphinx-rtd-theme"]
+    typing = ["mypy", "types-requests", {path = ".", extras = ["types"], only_deps = true}]
+    typing-test = [{include = "typing"}, {include = "test"}, "useful-types"]
+
+    [project.optional-dependencies]
+    types = ["typing-extensions"]
+
+Note how ``test`` and ``typing`` refer to the current package while ``docs``
+does not. This reflects the ability of Dependency Groups to be used in the same
+manner as extras, adding to dependencies, or completely independently.
+
+``typing-test`` is defined as a union of two existing groups, plus an
+additional package. ``typing`` includes an extra, ``types``, and this is
+included by extension under ``typing-test``.
+Under ``typing-test`` implementations may choose whether or not to use an
+editable installation of the current package or not, but they MUST treat
+``only_deps`` as ``false``.
+
+Package Building
+----------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+Use of Dependency Groups
+------------------------
+
+Tools which support Dependency Groups are expected to provide new options and
+interfaces to allow users to install from Dependency Groups.
+
+No syntax is defined for expressing the dependency group of a package, for two
+reasons:
+
+* it would not be valid to refer to the dependency groups of a third-party
+  package from pypi (because the data is defined to be unpublished)
+
+* there is not guaranteed to be a current package for dependency groups -- part
+  of their purpose is to support non-package projects
+
+For example, a possible pip interface for installing dependency groups
+would be:
+
+.. code:: shell
+
+    pip install --dependency-groups=test,typing
```

Comment:
Could this document also include examples of combining this with `--constraint` which is how pins/lockfiles would env matrixes would practically be defined (perhaps, generated with pip-tools, for example).

Reply:
Like @ofek I’m a strong -1 on this feature being used for locking. Specifically, dependency groups should never contain transitive dependencies, and so even with constraint files, won’t explicitly record the exact list of things to install.

An example of using dependency groups for something that works a bit like a lockfile would be misleading as to the intended use of the feature, IMO.

Reply:
My expectation is that tools like `pip-tools` would grow the ability to turn dependency groups into lockfiles. So I could definitely add an example of that.

Perhaps the best interface to document with respect to dependency locking is something like Poetry or PDM. These tools take more ownership of the file layout than pip-tools, so they know where the lockfile data is stored. That allows commands like
```
$TOOL install-group test
```
which implicitly reads the locked dependency group data, rather than pyproject.toml

Reply:
@pfmoore to be clear, I didn't mean to imply that the dependency groups should contain pinned or transitive deps. I rather expected something like `pip install --dependency-groups=test,typing --constraint=pip-tools-made-test-n-typing-lock.txt`.

@sirosen yes! That's my understanding, too — `pip-tools` should be taught to understand how to retrieve the dependency groups. As it's already able to get extras and extract the build deps, there already are mechanisms in place that would enable it to grab deps from another source relatively easily.

I don't think the PEP should advertise any lockfile layout, though. Not until there's a separate PEP that defines it.

Reply:
Let's keep non-editorial discussion separate from this PR, and in the discuss.python.org thread for this PR please.